### ISSUE TITLE
[Bug]: npm_translate_lock trying to download the npmrc file from remote downloader

### ISSUE THREAD

**Issue by gregjacobs** [aspect-build/rules_js#916]
### What happened?

Just trying rules_js@1.20.1, and `npm_translate_lock()` is attempting to "download" my local `//:.npmrc` file through our `--experimental_remote_downloader` service (which obviously won't work since the remote downloader service doesn't have my local files).

Error output:

```
(13:01:29) INFO: Repository npm instantiated at:
  /Users/user/project/WORKSPACE:71:19: in <toplevel>
  /private/var/tmp/_bazel_user/878fe649e7bc16de8f9e12289c1832cd/external/aspect_rules_js/npm/npm_import.bzl:434:24: in npm_translate_lock
Repository rule npm_translate_lock defined at:
  /private/var/tmp/_bazel_user/878fe649e7bc16de8f9e12289c1832cd/external/aspect_rules_js/npm/private/npm_translate_lock.bzl:101:37: in <toplevel>
(13:01:29) ERROR: An error occurred during the fetch of repository 'npm':
   Traceback (most recent call last):
	File "/private/var/tmp/_bazel_user/878fe649e7bc16de8f9e12289c1832cd/external/aspect_rules_js/npm/private/npm_translate_lock.bzl", line 57, column 41, in _impl
		state = npm_translate_lock_state.new(rctx)
	File "/private/var/tmp/_bazel_user/878fe649e7bc16de8f9e12289c1832cd/external/aspect_rules_js/npm/private/npm_translate_lock_state.bzl", line 538, column 10, in _new
		_init(priv, rctx, label_store)
	File "/private/var/tmp/_bazel_user/878fe649e7bc16de8f9e12289c1832cd/external/aspect_rules_js/npm/private/npm_translate_lock_state.bzl", line 72, column 29, in _init
		_copy_common_input_files(priv, rctx, label_store, pnpm_lock_exists)
	File "/private/var/tmp/_bazel_user/878fe649e7bc16de8f9e12289c1832cd/external/aspect_rules_js/npm/private/npm_translate_lock_state.bzl", line 265, column 29, in _copy_common_input_files
		_copy_input_file(priv, rctx, label_store, k)
	File "/private/var/tmp/_bazel_user/878fe649e7bc16de8f9e12289c1832cd/external/aspect_rules_js/npm/private/npm_translate_lock_state.bzl", line 415, column 18, in _copy_input_file
		rctx.download(
Error in download: java.io.IOException: io.grpc.StatusRuntimeException: UNAVAILABLE: asset not available: file:/Users/user/project/.npmrc
```

Traced it down to this piece of code:

```
def _copy_input_file(priv, rctx, label_store, key):
    # ...

    # use rctx.download to copy the file instead of rctx.read + rctx.file so that
    # binary files are handled correctly
    rctx.download(
        output = label_store.repository_path(key),
        url = "file:" + label_store.path(key),
    )
```

Is there another way to handle binary files without going through the remote downloader?

### Version

Development (host) and target OS/architectures: MacOS 12.16.3

Output of `bazel --version`: 6.0.0

Version of the Aspect rules, or other relevant rules from your
`WORKSPACE` or `MODULE.bazel` file: rules_js 1.20.1

### Any other information?

Unfortunately the implementation of our remote downloader is internal at our firm so I can't point you there, but its job is to populate the `--remote_cache` with assets and dependencies from the outside internet (such as rules_js itself) that are not directly accessible from our workstations.

I guess a generalized description of the problem is that the `npmrc` attribute of `npm_translate_lock` will break if `--experimental_remote_downloader` is set.

### Fund our work

- [ ] Sponsor our open source work by donating a [bug bounty](https://opencollective.com/aspect-build/)

**Comment by gregmagolan** [comment]
The way I've done it previously is with rctx.execute and `mkdir -p` and `cp` (plus different code paths for windows for `mkdir` and `xcopy`). That code is still in the old `npm_install` & `yarn_install` rules: https://github.com/bazelbuild/rules_nodejs/blob/ad43cf0ff7dea0ef8ea6c32016732b7cbb80f96a/internal/npm_install/npm_install.bzl#L554.

The `rctx.download` with `file:` was a nice clean way to avoid that. Reading the docs on https://bazel.build/reference/command-line-reference#flag--experimental_remote_downloader it seems like it should ignore `file:` urls. Does setting `--experimental_remote_downloader_local_fallback` resolve?

**Comment by gregjacobs** [comment]
Unfortunately, even with setting `--experimental_remote_downloader_local_fallback`, it seems tons of local file in rules_js@1.21.0 are trying to go through the `--experimental_remote_downloader`, including (seemingly) local repository files when (I guess) they're going to be consumed.

I get long warning messages with stack traces like this for files before I get failures shortly after for those same files not being able to be found:

```
Remote Cache: UNAVAILABLE: asset not available: file:/private/var/tmp/_bazel_user/878fe649e7bc16de8f9e12289c1832cd/external/aspect_rules_jest/jest/private/v29.3.1/pnpm-lock.yaml
io.grpc.StatusRuntimeException: UNAVAILABLE: asset not available:  file:/private/var/tmp/_bazel_user/878fe649e7bc16de8f9e12289c1832cd/external/aspect_rules_jest/jest/private/v29.3.1/pnpm-lock.yaml
	at io.grpc.stub.ClientCalls.toStatusRuntimeException(ClientCalls.java:262)
	at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:243)
	at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:156)
	at build.bazel.remote.asset.v1.FetchGrpc$FetchBlockingStub.fetchBlob(FetchGrpc.java:396)
	at com.google.devtools.build.lib.remote.downloader.GrpcRemoteDownloader.lambda$download$0(GrpcRemoteDownloader.java:135)
	at com.google.devtools.build.lib.remote.ReferenceCountedChannel.lambda$withChannelBlocking$2(ReferenceCountedChannel.java:85)
	at com.google.devtools.build.lib.remote.ReferenceCountedChannel.lambda$withChannel$4(ReferenceCountedChannel.java:108)
	at io.reactivex.rxjava3.internal.operators.single.SingleUsing.subscribeActual(SingleUsing.java:59)
	at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4855)
	at io.reactivex.rxjava3.internal.operators.single.SingleFlatMap$SingleFlatMapCallback.onSuccess(SingleFlatMap.java:85)
	at io.reactivex.rxjava3.internal.operators.single.SingleFlatMap$SingleFlatMapCallback$FlatMapSingleObserver.onSuccess(SingleFlatMap.java:112)
	at io.reactivex.rxjava3.internal.operators.single.SingleMap$MapSingleObserver.onSuccess(SingleMap.java:65)
	at io.reactivex.rxjava3.internal.operators.single.SingleDoOnDispose$DoOnDisposeObserver.onSuccess(SingleDoOnDispose.java:84)
	at io.reactivex.rxjava3.internal.operators.single.SingleDoOnError$DoOnError.onSuccess(SingleDoOnError.java:52)
	at io.reactivex.rxjava3.internal.operators.observable.ObservableSingleSingle$SingleElementObserver.onComplete(ObservableSingleSingle.java:110)
	at io.reactivex.rxjava3.internal.observers.DeferredScalarDisposable.complete(DeferredScalarDisposable.java:85)
	at io.reactivex.rxjava3.subjects.AsyncSubject.subscribeActual(AsyncSubject.java:233)
	at io.reactivex.rxjava3.core.Observable.subscribe(Observable.java:13176)
	at io.reactivex.rxjava3.internal.operators.observable.ObservableSingleSingle.subscribeActual(ObservableSingleSingle.java:36)
	at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4855)
	at io.reactivex.rxjava3.internal.operators.single.SingleDoOnError.subscribeActual(SingleDoOnError.java:35)
	at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4855)
	at io.reactivex.rxjava3.internal.operators.single.SingleDoOnDispose.subscribeActual(SingleDoOnDispose.java:38)
	at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4855)
	at io.reactivex.rxjava3.internal.operators.single.SingleMap.subscribeActual(SingleMap.java:35)
	at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4855)
	at io.reactivex.rxjava3.internal.operators.single.SingleFlatMap$SingleFlatMapCallback.onSuccess(SingleFlatMap.java:85)
	at io.reactivex.rxjava3.internal.operators.single.SingleCreate$Emitter.onSuccess(SingleCreate.java:68)
	at com.google.devtools.build.lib.remote.grpc.TokenBucket$1.onNext(TokenBucket.java:79)
	at io.reactivex.rxjava3.internal.util.NotificationLite.accept(NotificationLite.java:247)
	at io.reactivex.rxjava3.subjects.BehaviorSubject$BehaviorDisposable.test(BehaviorSubject.java:507)
	at io.reactivex.rxjava3.subjects.BehaviorSubject$BehaviorDisposable.emitFirst(BehaviorSubject.java:468)
	at io.reactivex.rxjava3.subjects.BehaviorSubject.subscribeActual(BehaviorSubject.java:224)
	at io.reactivex.rxjava3.core.Observable.subscribe(Observable.java:13176)
	at com.google.devtools.build.lib.remote.grpc.TokenBucket.lambda$acquireToken$0(TokenBucket.java:64)
	at io.reactivex.rxjava3.internal.operators.single.SingleCreate.subscribeActual(SingleCreate.java:40)
	at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4855)
	at io.reactivex.rxjava3.internal.operators.single.SingleFlatMap.subscribeActual(SingleFlatMap.java:37)
	at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4855)
	at io.reactivex.rxjava3.internal.operators.single.SingleDefer.subscribeActual(SingleDefer.java:43)
	at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4855)
	at io.reactivex.rxjava3.internal.operators.single.SingleFlatMap.subscribeActual(SingleFlatMap.java:37)
	at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4855)
	at io.reactivex.rxjava3.core.Single.blockingGet(Single.java:3644)
	at com.google.devtools.build.lib.remote.ReferenceCountedChannel.withChannelBlocking(ReferenceCountedChannel.java:85)
	at com.google.devtools.build.lib.remote.downloader.GrpcRemoteDownloader.lambda$download$1(GrpcRemoteDownloader.java:132)
	at com.google.devtools.build.lib.remote.Retrier.execute(Retrier.java:244)
	at com.google.devtools.build.lib.remote.RemoteRetrier.execute(RemoteRetrier.java:125)
	at com.google.devtools.build.lib.remote.RemoteRetrier.execute(RemoteRetrier.java:114)
	at com.google.devtools.build.lib.remote.downloader.GrpcRemoteDownloader.download(GrpcRemoteDownloader.java:130)
	at com.google.devtools.build.lib.bazel.repository.downloader.DelegatingDownloader.download(DelegatingDownloader.java:62)
	at com.google.devtools.build.lib.bazel.repository.downloader.DownloadManager.download(DownloadManager.java:258)
	at com.google.devtools.build.lib.bazel.repository.starlark.StarlarkBaseExternalContext.download(StarlarkBaseExternalContext.java:484)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at net.starlark.java.eval.MethodDescriptor.call(MethodDescriptor.java:162)
	at net.starlark.java.eval.BuiltinFunction.fastcall(BuiltinFunction.java:77)
	at net.starlark.java.eval.Starlark.fastcall(Starlark.java:638)
	at net.starlark.java.eval.Eval.evalCall(Eval.java:682)
	at net.starlark.java.eval.Eval.eval(Eval.java:497)
	at net.starlark.java.eval.Eval.exec(Eval.java:271)
	at net.starlark.java.eval.Eval.execStatements(Eval.java:82)
	at net.starlark.java.eval.Eval.execFunctionBody(Eval.java:66)
	at net.starlark.java.eval.StarlarkFunction.fastcall(StarlarkFunction.java:173)
	at net.starlark.java.eval.Starlark.fastcall(Starlark.java:638)
	at net.starlark.java.eval.Eval.evalCall(Eval.java:682)
	at net.starlark.java.eval.Eval.eval(Eval.java:497)
	at net.starlark.java.eval.Eval.exec(Eval.java:271)
	at net.starlark.java.eval.Eval.execStatements(Eval.java:82)
	at net.starlark.java.eval.Eval.execIf(Eval.java:205)
	at net.starlark.java.eval.Eval.exec(Eval.java:283)
	at net.starlark.java.eval.Eval.execStatements(Eval.java:82)
	at net.starlark.java.eval.Eval.execFor(Eval.java:126)
	at net.starlark.java.eval.Eval.exec(Eval.java:276)
	at net.starlark.java.eval.Eval.execStatements(Eval.java:82)
	at net.starlark.java.eval.Eval.execFunctionBody(Eval.java:66)
	at net.starlark.java.eval.StarlarkFunction.fastcall(StarlarkFunction.java:173)
	at net.starlark.java.eval.Starlark.fastcall(Starlark.java:638)
	at net.starlark.java.eval.Eval.evalCall(Eval.java:682)
	at net.starlark.java.eval.Eval.eval(Eval.java:497)
	at net.starlark.java.eval.Eval.exec(Eval.java:271)
	at net.starlark.java.eval.Eval.execStatements(Eval.java:82)
	at net.starlark.java.eval.Eval.execFunctionBody(Eval.java:66)
	at net.starlark.java.eval.StarlarkFunction.fastcall(StarlarkFunction.java:173)
	at net.starlark.java.eval.Starlark.fastcall(Starlark.java:638)
	at net.starlark.java.eval.Eval.evalCall(Eval.java:682)
	at net.starlark.java.eval.Eval.eval(Eval.java:497)
	at net.starlark.java.eval.Eval.exec(Eval.java:271)
	at net.starlark.java.eval.Eval.execStatements(Eval.java:82)
	at net.starlark.java.eval.Eval.execFunctionBody(Eval.java:66)
	at net.starlark.java.eval.StarlarkFunction.fastcall(StarlarkFunction.java:173)
	at net.starlark.java.eval.Starlark.fastcall(Starlark.java:638)
	at net.starlark.java.eval.Eval.evalCall(Eval.java:682)
	at net.starlark.java.eval.Eval.eval(Eval.java:497)
	at net.starlark.java.eval.Eval.execAssignment(Eval.java:109)
	at net.starlark.java.eval.Eval.exec(Eval.java:268)
	at net.starlark.java.eval.Eval.execStatements(Eval.java:82)
	at net.starlark.java.eval.Eval.execFunctionBody(Eval.java:66)
	at net.starlark.java.eval.StarlarkFunction.fastcall(StarlarkFunction.java:173)
	at net.starlark.java.eval.Starlark.fastcall(Starlark.java:638)
	at net.starlark.java.eval.Starlark.call(Starlark.java:604)
	at com.google.devtools.build.lib.bazel.repository.starlark.StarlarkRepositoryFunction.fetch(StarlarkRepositoryFunction.java:220)
	at com.google.devtools.build.lib.rules.repository.RepositoryDelegatorFunction.fetchRepository(RepositoryDelegatorFunction.java:413)
	at com.google.devtools.build.lib.rules.repository.RepositoryDelegatorFunction.compute(RepositoryDelegatorFunction.java:344)
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:571)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:382)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

**Comment by gregjacobs** [comment]
I'm going to close this issue here - this seems like a Bazel bug. Have raised https://github.com/bazelbuild/bazel/issues/17771
